src/Components/Web.JS/src/Platform/WebAssemblyResourceLoader.ts

```diff
@@ -0,0 +1,195 @@
+    // We want to keep the cache small because, even though the browser will evict entries if it
+    // gets too big, we don't want to be considered problematic by the end user viewing storage stats
+    const cachedRequests = await this.cache.keys();
+    const deletionPromises = cachedRequests.map(async cachedRequest => {
+      if (!(cachedRequest.url in this.usedCacheKeys)) {
+        await this.cache.delete(cachedRequest);
+      }
+    });
+
+    return Promise.all(deletionPromises);
+  }
+
+  private async tryValidateAndCacheAsync(name: string, cacheKey: string, expectedContentHash: string, networkResponse: Response): Promise<ArrayBuffer> {
+    const responseBuffer = await networkResponse.arrayBuffer();
+
+    // Now is an ideal moment to capture the performance stats for the request, since it
+    // only just completed and is most likely to still be in the buffer. However this is
+    // only done on a 'best effort' basis. Even if we do receive an entry, some of its
+    // properties may be blanked out if it was a CORS request.
+    const performanceEntry = getPerformanceEntry(networkResponse.url);
+    const responseBytes = (performanceEntry && performanceEntry.encodedBodySize) || undefined;
+    this.networkLoads[name] = { responseBytes };
+
+    // crypto.subtle is only enabled on localhost or HTTPS origins
+    // We only write to the cache if we can validate the content hashes
+    if (typeof crypto !== 'undefined' && !!crypto.subtle) {
+      await assertContentHashMatchesAsync(name, responseBuffer, expectedContentHash);
+
+      // Build a custom response object so we can track extra data such as responseBytes
+      // We can't rely on the server sending content-length (ASP.NET Core doesn't by default)
+      await this.cache.put(cacheKey, new Response(responseBuffer, {
+        headers: {
+          'content-type': networkResponse.headers.get('content-type') || '',
+          'content-length': (responseBytes || networkResponse.headers.get('content-length') || '').toString()
+        }
+      }));
+    }
+
+    return responseBuffer;
+  }
+}
+
+function countTotalBytes(loads: LoadLogEntry[]) {
+  return loads.reduce((prev, item) => prev + (item.responseBytes || 0), 0);
+}
+
+function toDataSizeString(byteCount: number) {
+  return `${(byteCount / (1024*1024)).toFixed(2)} MB`;
+}
+
+function getPerformanceEntry(url: string): PerformanceResourceTiming | undefined {
+  if (typeof performance !== 'undefined') {
+    return performance.getEntriesByName(url)[0] as PerformanceResourceTiming;
+  }
+}
+
+async function assertContentHashMatchesAsync(name: string, data: ArrayBuffer, expectedHashPrefix: string) {
+  const actualHashBuffer = await crypto.subtle.digest('SHA-256', data);
+  const actualHash = new Uint8Array(actualHashBuffer);
+  for (var byteIndex = 0; byteIndex*2 < expectedHashPrefix.length; byteIndex++) {
+    const expectedByte = parseInt(expectedHashPrefix.substr(byteIndex * 2, 2), 16);
```

Comment:
Can we change the hashing format to be base64 instead of hex? It's way more standard on the web and allows us to build the entries as data urls that can be trivially read on the client (and makes the validation straightforward)
`data:text/plain;base64,<<urlEncodedHash>>`
`await fetch('data:text/plain;base64,<<urlEncodedHash>>').then(r => r.arrayBuffer()));`
and at that point is just a byte by byte comparison rather than rolling our own logic.

Reply:
Yes, I've now changed the hashes to be base64 and compared in the way you suggested.

This was a great idea - thanks!

Reply:
As I'm sure you've realised, another massive payoff from this suggestion is that now we can use `{ integrity: ... }` on the `fetch`. This definitely feels like we're aligning with how things are meant to be done. So, thanks again for the suggestion!